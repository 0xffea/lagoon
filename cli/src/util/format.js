// @flow

import R from 'ramda';
import {
  FORMAT,
  FORMAT_CHOICE_TABLE,
  FORMAT_CHOICE_SIMPLE,
  FORMAT_CHOICE_JSON,
  FORMAT_CHOICE_CSV,
} from '../config/globalOptions';
import { getConfig } from '../config';

import type { FormatData } from '../types/Format';

export default function format(
  data: FormatData,
  options: ?{
    // Optional transforms for data
    transforms?: { [key: string]: (FormatData) => FormatData },
  },
) {
  const config = getConfig();
  const configuredFormat = R.prop(FORMAT, config);
  const transformByFormat =
    R.path(['transforms', configuredFormat], options) ||
    // Default to just returning the data
    R.identity;
  const transformedData = transformByFormat(data);

  return R.cond([
    [
      R.equals(FORMAT_CHOICE_TABLE),
      R.always(require('../formatters/table').default(transformedData)),
    ],
    [
      R.equals(FORMAT_CHOICE_SIMPLE),
      R.always(require('../formatters/simple').default(transformedData)),
    ],
    [
      R.equals(FORMAT_CHOICE_JSON),
      R.always(require('../formatters/json').default(transformedData)),
    ],
    [
      R.equals(FORMAT_CHOICE_CSV),
      R.always(require('../formatters/csv').default(transformedData)),
    ],
  ])(configuredFormat);
}
