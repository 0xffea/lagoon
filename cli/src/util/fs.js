// @flow

/**
 * This module exists to promisify all the fs functionality
 */

import fs from 'fs';

import type { Stats } from 'fs';

export function readFile(filename: string, enc?: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fs.readFile(filename, enc || 'utf8', (err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(data);
      }
    });
  });
}

export function writeFile(
  filename: string,
  data: Buffer | string,
  options?: Object | string,
): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.writeFile(filename, data, options, (err) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export function deleteFile(filename: string): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.unlink(filename, (err) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export function lstat(path: string): Promise<Stats> {
  return new Promise((resolve, reject) => {
    fs.lstat(path, (err, stat) => {
      if (err) {
        reject(err);
      } else {
        resolve(stat);
      }
    });
  });
}

export async function doesFileExist(file: string): Promise<boolean> {
  try {
    const stats = await lstat(file);

    if (stats.isFile()) {
      return true;
    }

    return false;
  } catch (err) {
    return false;
  }
}

export const statSync = fs.statSync;
