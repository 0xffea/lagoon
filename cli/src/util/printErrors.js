// @flow

import { red } from 'chalk';
import R from 'ramda';

type Cerr = typeof console.error;
type LaguError = { message: string };

export function printErrors(cerr: Cerr, ...errors: Array<LaguError>): number {
  R.compose(
    R.forEach(cerr),
    R.map(red),
    R.map(R.prop('message')),
  )(errors);
  return 1;
}

export function printNoConfigError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      '.lagoon.yml config file not found. Please create one with "lagoon init".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printProjectConfigurationError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      'No project configured. Please create a .lagoon.yml config file with "lagoon init" or pass a project to this command via the --project option.\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printGraphQLErrors(
  cerr: Cerr,
  ...errors: Array<LaguError>
): number {
  if (
    R.find(R.propEq('message', 'Unauthorized - Bearer Token Required'))(errors)
  ) {
    return printErrors(cerr, {
      message:
        'No authentication token found. Please log in first with "lagoon login".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-login',
    });
  }
  const errorMessage =
    R.length(errors) === 1
      ? 'Oops! The lagoon API returned an error:'
      : 'Oops! The lagoon API returned errors:';
  return printErrors(cerr, { message: errorMessage }, ...errors);
}
