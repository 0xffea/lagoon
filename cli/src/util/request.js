// @flow

import defer from './defer';

export type RequestOptions = {
  hostname: string,
  path?: string,
  port?: number | string,
  method: 'POST' | 'GET',
  headers?: Object,
  rejectUnauthorized?: boolean,
  body?: string,
};

export default function request(options: RequestOptions): Promise<Object> {
  const { body, port = 443 } = options;

  const def = defer();

  const thing = Number(port) === 80 ? require('http') : require('https');

  const req = thing.request(options, res => {
    res.setEncoding('utf8');

    let rawData = '';
    res.on('data', chunk => {
      rawData += chunk;
    });
    res.on('end', () => {
      try {
        let parsed = JSON.parse(rawData);
        def.resolve(parsed);
      }
      catch (e) {
        def.reject(e);
      }
    });
  });

  if (body) {
    req.write(body);
  }

  req.end();

  req.on('error', e => def.reject(e));

  return def.promise;
}
