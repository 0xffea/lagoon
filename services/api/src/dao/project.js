// @flow

const R = require('ramda');
const logger = require('../logger');
const {
  ifNotAdmin,
  inClauseOr,
  knex,
  prepare,
  query,
  whereAnd,
  isPatchEmpty,
} = require('./utils');

// This contains the sql query generation logic
const Sql = {
  updateProject: ({ permissions: { projects } }, { id, patch }) =>
    knex('project')
      .where('id', '=', id)
      .whereIn('id', projects)
      .update(patch)
      .toString(),
  selectProject: (id /* : number */) =>
    knex('project')
      .where('id', id)
      .toString(),
  selectProjectIdByName: name =>
    knex('project')
      .where('name', name)
      .select('id')
      .toString(),
  selectProjectIdsByCustomerIds: customerIds =>
    knex('project')
      .select('id')
      .whereIn('customer', customerIds)
      .toString(),
  selectCustomer: id =>
    knex('customer')
      .where('id', id)
      .toString(),
  truncateProject: () =>
    knex('project')
      .truncate()
      .toString(),
};

const Helpers = {
  getProjectById: async (sqlClient, id) => {
    const rows = await query(sqlClient, Sql.selectProject(id));
    return R.prop(0, rows);
  },
  getProjectIdByName: async (sqlClient, name) => {
    const pidResult = await query(sqlClient, Sql.selectProjectIdByName(name));

    const amount = R.length(pidResult);
    if (amount > 1) {
      throw new Error(
        `Multiple project candidates for '${name}' (${amount} found). Do nothing.`,
      );
    }

    if (amount === 0) {
      throw new Error(`Not found: '${name}'`);
    }

    const pid = R.path(['0', 'id'], pidResult);

    return pid;
  },
  getProjectIdsByCustomerIds: async (sqlClient, customerIds) =>
    query(sqlClient, Sql.selectProjectIdsByCustomerIds(customerIds)),
  getCustomerByCustomerId: async (sqlClient, id) => {
    const rows = await query(sqlClient, Sql.selectCustomer(id));
    return R.prop(0, rows);
  },
};

const getAllProjects = ({ sqlClient }) => async (cred, args) => {
  const { customers, projects } = cred.permissions;
  // We need one "WHERE" keyword, but we have multiple optional conditions
  const where = whereAnd([
    args.createdAfter ? 'created >= :created_after' : '',
    args.gitUrl ? 'git_url = :git_url' : '',
    ifNotAdmin(
      cred.role,
      `(${inClauseOr([['customer', customers], ['project.id', projects]])})`,
    ),
  ]);

  const prep = prepare(sqlClient, `SELECT * FROM project ${where}`);
  const rows = await query(sqlClient, prep(args));

  return rows;
};

const getProjectByEnvironmentId = ({ sqlClient }) => async (cred, eid) => {
  const { customers, projects } = cred.permissions;
  const prep = prepare(
    sqlClient,
    `SELECT
        p.*
      FROM environment e
      JOIN project p ON e.project = p.id
      WHERE e.id = :eid
      ${ifNotAdmin(
    cred.role,
    `AND (${inClauseOr([['p.customer', customers], ['p.id', projects]])})`,
  )}
      LIMIT 1
    `,
  );

  const rows = await query(sqlClient, prep({ eid }));

  return rows ? rows[0] : null;
};

const getProjectByGitUrl = ({ sqlClient }) => async (cred, args) => {
  const { customers, projects } = cred.permissions;
  const str = `
      SELECT
        *
      FROM project
      WHERE git_url = :git_url
      ${ifNotAdmin(
    cred.role,
    `AND (${inClauseOr([
      ['customer', customers],
      ['project.id', projects],
    ])})`,
  )}
      LIMIT 1
    `;

  const prep = prepare(sqlClient, str);
  const rows = await query(sqlClient, prep(args));

  return rows ? rows[0] : null;
};

const getProjectByName = ({ sqlClient }) => async (cred, args) => {
  const { customers, projects } = cred.permissions;
  const str = `
      SELECT
        *
      FROM project
      WHERE name = :name
      ${ifNotAdmin(
    cred.role,
    `AND (${inClauseOr([
      ['customer', customers],
      ['project.id', projects],
    ])})`,
  )}
    `;

  const prep = prepare(sqlClient, str);

  const rows = await query(sqlClient, prep(args));
  return rows[0];
};

const addProject = ({
  sqlClient,
  keycloakClient,
  searchguardClient,
  kibanaClient,
}) => async (cred, input) => {
  const { customers } = cred.permissions;
  const cid = input.customer.toString();

  if (cred.role !== 'admin' && !R.contains(cid, customers)) {
    throw new Error('Project creation unauthorized.');
  }

  const prep = prepare(
    sqlClient,
    `CALL CreateProject(
        :id,
        :name,
        :customer,
        :git_url,
        ${input.subfolder ? ':subfolder' : 'NULL'},
        :openshift,
        ${
  input.openshiftProjectPattern ? ':openshift_project_pattern' : 'NULL'
},
        ${
  input.activeSystemsDeploy
    ? ':active_systems_deploy'
    : '"lagoon_openshiftBuildDeploy"'
},
        ${
  input.activeSystemsPromote
    ? ':active_systems_promote'
    : '"lagoon_openshiftBuildDeploy"'
},
        ${
  input.activeSystemsRemove
    ? ':active_systems_remove'
    : '"lagoon_openshiftRemove"'
},
        ${input.branches ? ':branches' : '"true"'},
        ${input.pullrequests ? ':pullrequests' : '"true"'},
        ${input.productionEnvironment ? ':production_environment' : 'NULL'},
        ${input.autoIdle ? ':auto_idle' : '1'},
        ${input.storageCalc ? ':storage_calc' : '1'}
      );
    `,
  );

  const rows = await query(sqlClient, prep(input));
  const project = R.path([0, 0], rows);

  try {
    // Create a group in Keycloak named the same as the project
    await keycloakClient.groups.create({
      name: R.prop('name', project),
    });
  } catch (err) {
    if (err.response.status === 409) {
      logger.warn(
        `Failed to create already existing Keycloak group "${R.prop(
          'name',
          project,
        )}"`,
      );
    } else {
      logger.error(`SearchGuard create role error: ${err}`);
      throw new Error(`SearchGuard create role error: ${err}`);
    }
  }

  const customer = await Helpers.getCustomerByCustomerId(
    sqlClient,
    project.customer,
  );

  try {
    // Create a new SearchGuard Role for this project with the same name as the Project
    await searchguardClient.put(`roles/${project.name}`, {
      body: {
        indices: {
          [`*-${project.name}-*`]: {
            '*': ['READ'],
          },
        },
        tenants: {
          [customer.name]: 'RW',
        },
      },
    });
  } catch (err) {
    logger.error(`SearchGuard create role error: ${err}`);
    throw new Error(`SearchGuard create role error: ${err}`);
  }

  // Create index-patterns for this project
  for (const log of [
    'application-logs',
    'router-logs',
    'container-logs',
    'lagoon-logs',
  ]) {
    try {
      await kibanaClient.post(
        `saved_objects/index-pattern/${log}-${project.name}-*`,
        {
          body: {
            attributes: {
              title: `${log}-${project.name}-*`,
            },
          },
          headers: {
            sgtenant: customer.name,
          },
        },
      );
    } catch (err) {
      // 409 Errors are expected and mean that there is already an index-pattern with that name defined, we ignore them
      if (err.statusCode !== 409) {
        logger.error(
          `Kibana Error during setup of index pattern ${log}-${
            project.name
          }-*: ${err}`,
        );
        // Don't fail if we have Kibana Errors, as they are "non-critical"
      }
    }
  }

  try {
    const currentSettings = await kibanaClient.get('kibana/settings', {
      headers: {
        sgtenant: customer.name,
      },
    });

    // Define a default Index if there is none yet
    if (!currentSettings.body.settings.defaultIndex) {
      await kibanaClient.post('kibana/settings', {
        body: {
          changes: {
            defaultIndex: `container-logs-${project.name}-*`,
            'telemetry:optIn': false, // also opt out of telemetry from xpack
          },
        },
        headers: {
          sgtenant: customer.name,
        },
      });
    }
  } catch (err) {
    logger.error(`Kibana Error during config of default Index: ${err}`);
    // Don't fail if we have Kibana Errors, as they are "non-critical"
  }

  return project;
};

const deleteProject = ({ sqlClient, searchguardClient }) => async (
  cred,
  input,
) => {
  const { projects } = cred.permissions;

  // Will throw on invalid conditions
  const pid = await Helpers.getProjectIdByName(sqlClient, input.project);

  if (cred.role !== 'admin') {
    if (!R.contains(pid, projects)) {
      throw new Error('Unauthorized.');
    }
  }

  const prep = prepare(sqlClient, 'CALL DeleteProject(:project)');
  await query(sqlClient, prep(input));

  try {
    // Delete SearchGuard Role for this project with the same name as the Project
    await searchguardClient.delete(`roles/${input.project}`);
  } catch (err) {
    logger.error(`SearchGuard delete role error: ${err}`);
    throw new Error(`SearchGuard delete role error: ${err}`);
  }
  // TODO: maybe check rows for changed result
  return 'success';
};

const updateProject = ({ sqlClient }) => async (cred, input) => {
  const { projects } = cred.permissions;
  const pid = input.id.toString();

  if (cred.role !== 'admin' && !R.contains(pid, projects)) {
    throw new Error('Unauthorized');
  }

  if (isPatchEmpty(input)) {
    throw new Error('input.patch requires at least 1 attribute');
  }

  await query(sqlClient, Sql.updateProject(cred, input));
  return Helpers.getProjectById(pid);
};

const deleteAllProjects = ({ sqlClient }) => async ({ role }) => {
  if (role !== 'admin') {
    throw new Error('Unauthorized.');
  }

  await query(sqlClient, Sql.truncateProject());

  // TODO: Check rows for success
  return 'success';
};

module.exports = {
  Sql,
  Queries: {
    deleteProject,
    addProject,
    getProjectByName,
    getProjectByGitUrl,
    getProjectByEnvironmentId,
    getAllProjects,
    updateProject,
    deleteAllProjects,
  },
  Helpers,
};
