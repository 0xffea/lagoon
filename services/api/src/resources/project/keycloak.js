// @flow

const R = require('ramda');
const keycloakClient = require('../../clients/keycloakClient');
const logger = require('../../logger');

const KeycloakOperations = {
  findGroupIdByName: async (name /* : string */) =>
    R.path(
      [0, 'id'],
      await keycloakClient.groups.find({
        search: name,
      }),
    ),
  deleteGroup: async (name /* : string */) => {
    try {
      // Find the Keycloak group id with the name
      const keycloakGroupId = await KeycloakOperations.findGroupIdByName(name);

      // Delete the group
      await keycloakClient.groups.del({ id: keycloakGroupId });

      logger.debug(`Deleted Keycloak group "${name}"`);
    } catch (err) {
      logger.error(`Error deleting Keycloak group: ${err}`);
      throw new Error(`Error deleting Keycloak group: ${err}`);
    }
  },
};

module.exports = KeycloakOperations;
