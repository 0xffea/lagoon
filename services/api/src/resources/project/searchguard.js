// @flow

const R = require('ramda');
const searchguardClient = require('../../clients/searchguardClient');
const kibanaClient = require('../../clients/kibanaClient');
const logger = require('../../logger');

const { getCustomerById } = require('../customer/helpers');

const SearchguardOperations = {
  addProject: async (project /* : any */) => {
    const customer = await getCustomerById(project.customer);

    try {
    // Create a new SearchGuard Role for this project with the same name as the Project
      await searchguardClient.put(`roles/${project.name}`, {
        body: {
          indices: {
            [`*-${project.name}-*`]: {
              '*': ['READ'],
            },
          },
          tenants: {
            [customer.name]: 'RW',
          },
        },
      });
      logger.debug(`Created SearchGuard role "${project.name}"`);
    } catch (err) {
      logger.error(`SearchGuard create role error: ${err}`);
      throw new Error(`SearchGuard create role error: ${err}`);
    }

    // Create index-patterns for this project

    for (const log of [
      'application-logs',
      'router-logs',
      'container-logs',
      'lagoon-logs',
    ]) {
      try {
        await kibanaClient.post(
          `saved_objects/index-pattern/${log}-${project.name}-*`,
          {
            body: {
              attributes: {
                title: `${log}-${project.name}-*`,
                timeFieldName: '@timestamp',
              },
            },
            headers: {
              sgtenant: customer.name,
            },
          },
        );
        logger.debug(`Created index-pattern "${log}-${project.name}-*"`);
      } catch (err) {
      // 409 Errors are expected and mean that there is already an index-pattern with that name defined, we ignore them
        if (err.statusCode !== 409) {
          logger.error(
            `Kibana Error during setup of index pattern ${log}-${
              project.name
            }-*: ${err}`,
          );
        // Don't fail if we have Kibana Errors, as they are "non-critical"
        } else {
          logger.debug(`index-pattern "${log}-${project.name}-*" already existing`);
        }
      }
    }

    try {
      const currentSettings = await kibanaClient.get('kibana/settings', {
        headers: {
          sgtenant: customer.name,
        },
      });
      // Define a default Index if there is none yet
      if (!('defaultIndex' in currentSettings.body.settings)) {
        await kibanaClient.post('kibana/settings', {
          body: {
            changes: {
              defaultIndex: `container-logs-${project.name}-*`,
              'telemetry:optIn': false, // also opt out of telemetry from xpack
            },
          },
          headers: {
            sgtenant: customer.name,
          },
        });
        logger.debug(`Configured default index for tenant "${customer.name}" to  "container-logs-${project.name}-*"`);
      } else {
        logger.debug(`Configured default index for tenant "${customer.name}" was already set to "${currentSettings.body.settings.defaultIndex.userValue}"`);
      }
    } catch (err) {
      logger.error(`Kibana Error during config of default Index: ${err}`);
      // Don't fail if we have Kibana Errors, as they are "non-critical"
    }
  },
};

module.exports = SearchguardOperations;
