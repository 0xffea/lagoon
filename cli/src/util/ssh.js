// @flow

import inquirer from 'inquirer';
import R from 'ramda';
import { Client } from 'ssh2';
import { fileExists } from '../util/fs';
import { printErrors } from '../printErrors';

type ExecOptions = {};

type Connection = {
  exec: (
    command: string,
    options?: ExecOptions,
    callback: Function,
  ) => Connection,
  on: (event: string, callback: Function) => Connection,
  end: () => Connection,
};

type ConnectArgs = {
  host: string,
  port: number,
  username: string,
  privateKey: string | Buffer,
  passphrase?: string,
};

export async function sshConnect(args: ConnectArgs): Promise<Connection> {
  return new Promise((resolve, reject) => {
    const connection = new Client();

    connection.on('ready', () => {
      resolve(connection);
    });

    connection.on('error', (error) => {
      reject(error);
    });

    connection.connect(args);
  });
}

export async function sshExec(
  connection: Connection,
  command: string,
  options?: ExecOptions = {},
): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    connection.exec(command, options, (error, stream) => {
      if (error) {
        reject(error);
      } else {
        stream.on('data', (data) => {
          resolve(data);
        });

        stream.stderr.on('data', (data) => {
          reject(new Error(data));
        });
      }
    });
  });
}

type GetPrivateKeyPathArgs = {
  fileExistsAtDefaultPath: boolean,
  defaultPrivateKeyPath: string,
  identityOption?: string,
  cerr: typeof console.error,
};

export const getPrivateKeyPath = async (
  args: GetPrivateKeyPathArgs,
): Promise<string> =>
  R.cond([
    // If the identity option for the command has been specified and the file at the path exists, use the value of that
    [
      R.propSatisfies(
        // Option is not null or undefined
        R.complement(R.isNil),
        'identityOption',
      ),
      R.prop('identityOption'),
    ],
    // If a file exists at the default private key path, use that
    [R.prop('fileExistsAtDefaultPath'), R.prop('defaultPrivateKeyPath')],
    // If none of the previous conditions have been satisfied, ask the user if they want to overwrite the file
    [
      R.T,
      async ({ cerr }) => {
        // TODO: Do now.
        // TODO: Ramdaify / FPify?
        let privateKeyPath = '';
        let privateKeyExists = false;
        while (!privateKeyPath || !privateKeyExists) {
          // eslint-disable-next-line no-await-in-loop
          const promptReturn = await inquirer.prompt([
            {
              type: 'input',
              name: 'privateKeyPath',
              message: 'Path to private key file',
            },
          ]);
          privateKeyPath = promptReturn.privateKeyPath;

          // eslint-disable-next-line no-await-in-loop
          privateKeyExists = await fileExists(privateKeyPath);
          if (!privateKeyExists) {
            printErrors(cerr, 'File does not exist at given path!');
          }
        }
        return privateKeyPath;
      },
    ],
  ])(args);

export const getPrivateKeyPassphrase = async (
  privateKeyHasEncryption: boolean,
): Promise<string> =>
  R.ifElse(
    // If the private key doesn't have encryption...
    R.not,
    // ... return an empty string
    R.always(''),
    // If the private key has encryption, ask for the password
    async () => {
      const { passphrase } = await inquirer.prompt([
        {
          type: 'password',
          name: 'passphrase',
          message: 'Private key password (never saved)',
          default: '',
        },
      ]);
      return passphrase;
    },
  )(privateKeyHasEncryption);
