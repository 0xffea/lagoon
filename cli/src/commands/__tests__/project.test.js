// @flow

import { queryGraphQL } from '../../util/queryGraphQL';
import { projectDetails } from '../project';

jest.mock('../../util/queryGraphQL');

const _mock = (mockFn: any): JestMockFn<any, any> => mockFn;

const mockErrorResponse = {
  errors: [{ message: 'something something error' }],
};

describe('projectDetails', () => {
  const mockResponse1 = {
    data: {
      projectByName: {
        name: 'credentialstest-project1',
        customer: {
          name: 'credentialtest-customer1',
        },
        git_url: 'project1.git',
        active_systems_deploy: 'lagoon_openshiftBuildDeploy',
        active_systems_remove: 'lagoon_openshiftRemove',
        branches: 'true',
        pullrequests: null,
        openshift: {
          name: 'credentialtest-openshift',
        },
        created: '2018-01-15 11:09:35',
      },
    },
  };

  it('should display error, if GraphQL sends error messages', async () => {
    _mock(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve(mockErrorResponse),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await projectDetails({
      clog,
      cerr,
      options: { project: 'some_project' },
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should show error on missing project', async () => {
    _mock(queryGraphQL).mockImplementationOnce(() => Promise.resolve({}));

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await projectDetails({
      clog,
      cerr,
      options: { project: 'not_existing' },
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should show details for given project', async () => {
    _mock(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve(mockResponse1),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await projectDetails({
      clog,
      cerr,
      options: { project: 'myproject' },
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
