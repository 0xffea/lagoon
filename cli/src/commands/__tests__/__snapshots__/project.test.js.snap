// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The Lagoon API returned an error:[39m",
  ],
  Array [
    "[31msomething something error[39m",
  ],
]
`;

exports[`handler should show details for given project 1`] = `
Array [
  Array [
    "Project details for 'myproject':",
  ],
  Array [
    "╔═══════════════════════╤═════════════════════════════╗
║ Name                  │ credentialstest-project1    ║
╟───────────────────────┼─────────────────────────────╢
║ Customer              │ credentialtest-customer1    ║
╟───────────────────────┼─────────────────────────────╢
║ Git URL               │ project1.git                ║
╟───────────────────────┼─────────────────────────────╢
║ Active Systems Deploy │ lagoon_openshiftBuildDeploy ║
╟───────────────────────┼─────────────────────────────╢
║ Active Systems Remove │ lagoon_openshiftRemove      ║
╟───────────────────────┼─────────────────────────────╢
║ Branches              │ true                        ║
╟───────────────────────┼─────────────────────────────╢
║ Pull Requests         │ null                        ║
╟───────────────────────┼─────────────────────────────╢
║ Openshift             │ credentialtest-openshift    ║
╟───────────────────────┼─────────────────────────────╢
║ Created               │ 2018-01-15 11:09:35         ║
╚═══════════════════════╧═════════════════════════════╝
",
  ],
]
`;

exports[`handler should show error on missing project 1`] = `
Array [
  Array [
    "No project 'not_existing' found.",
  ],
]
`;
